compiler/src/dotty/tools/dotc/core/tasty/TastyString.scala

```diff
@@ -0,0 +1,20 @@
+package dotty.tools.dotc.core.tasty
+
+/** Utils for String representation of TASTY */
+object TastyString {
+
+  /** Decode the TASTY String into TASTY bytes */
+  def stringToTasty(str: String): Array[Byte] = { // TODO factor out this and tastyToString
+    val bytes = new Array[Byte](str.length)
+    for (i <- str.indices) bytes(i) = str.charAt(i).toByte
+    bytes
+  }
+
+  /** Encode TASTY bytes into a TASTY String */
+  def tastyToString(bytes: Array[Byte]): String = {
```

Comment:
`new String(bytes)`?

Reply:
Same problem

Reply:
This is a tricky problem. Looking at Stackoverflow, people say you should use a Codec for this, typically Base64. The scheme of mapping all bytes to ranges 0..255 looks like it would work, but it's not optimal. Strings are represented in Classfiles as UTF8 characters, with one byte for ranges 0.127 and two bytes for ranges 128-255. This means that, assuming a uniform bit distribution you get an overhead of 50%. Doing a 8->7 bit codec would give an overhead of less than 15%. 

There's another problem of string size. Strings are limited to 65365 characters. This might not be enough for a larger quoted program. 

scalac solves both of these problems when serializing its pickles as annotations. I think we should copy that scheme. I tried to find it but could not. @retronym @lrytz @adriaanm does one of you have an idea where the code that serializes a Pickle as an annotation is?

Reply:
I think we can leave it like this for this PR, but then we should open an issue for future improvements.

Reply:
I will start looking at the alternatives. I also think we should start with this for now to unblock the next PRs and allow people to use it.

Reply:
Agreed.

Reply:
@odersky https://github.com/scala/scala/blob/2.13.x/src/compiler/scala/tools/nsc/backend/jvm/BCodeHelpers.scala#L1036-L1120

Reply:
@lrytz thanks for the link. Could you also point me to the place where the `String`/`Array[Strings]` are converted back into an `Array[Byte]`. Thanks.

Reply:
It took me a while to find it.. Need to clean this up / document. Method [`parseScalaSigBytes`](https://github.com/scala/scala/blob/68d66968eede247adf6d784e0d8f0f343d9124d4/src/compiler/scala/tools/nsc/symtab/classfile/ClassfileParser.scala#L1125) calls `ConstantPool.getBytes` which goes through `ByteCodecs.decode`.

The encoding is explained here http://www.scala-lang.org/old/sites/default/files/sids/dubochet/Mon,%202010-05-31,%2015:25/Storage%20of%20pickled%20Scala%20signatures%20in%20class%20files.pdf
  - first map all 8-bit bytes to 7 bits (shifting the rest)
  - then increment all by 1 (in 7 bits), so 0x7f becomes 0x00
  - then encode 0x00 as 0xc0 0x80, which is an overlong utf 8 encoding for zero. it's what the jvm classfile spec uses to avoid having 0x00 in strings. it's called "modified utf 8".

the reason for the incrementing by 1 that 0x7f is expected to be less common than 0x00, so the two byte encoding hits less often.

The confusing part is that the class `ScalaSigBytes` used in the backend to encode the signature uses `ByteCodecs.encode8to7`, but does the `+1` itself. It doesn't need to map 0x00 to the two byte version because ASM will do it when writing the annotation to the classfile. However, in the unpickler, we don't use ASM to read the annotation, but just get the bytes from the classfile directly. So there we'll see the two byte encoding. `ByteCodecs.decode` does that work.